UNIXFSv1.md

```diff
@@ -0,0 +1,487 @@
+# ![](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square) UnixFS  <!-- omit in toc -->
+
+**Author(s)**:
+- NA
+
+* * *
+
+**Abstract**
+
+UnixFS is a [protocol-buffers](https://developers.google.com/protocol-buffers/) based format for describing files, directories, and symlinks as merkle-dags in IPFS.
+
+Draft work and discussion on a specification for the upcoming version 2 of the UnixFS format is happening in the [`ipfs/unixfs-v2` repo](https://github.com/ipfs/unixfs-v2). Please see the issues there for discussion and PRs for drafts.
+
+## Table of Contents <!-- omit in toc -->
+
+- [Implementations](#implementations)
+- [Data Format](#data-format)
+- [Metadata](#metadata)
+	- [Deduplication and inlining](#deduplication-and-inlining)
+- [Importing](#importing)
+	- [Chunking](#chunking)
+	- [Layout](#layout)
+- [Exporting](#exporting)
+- [Design decision rationale](#design-decision-rationale)
+	- [Metadata](#metadata-1)
+		- [Separate Metadata node](#separate-metadata-node)
+		- [Metadata in the directory](#metadata-in-the-directory)
+		- [Metadata in the file](#metadata-in-the-file)
+		- [Side trees](#side-trees)
+		- [Side database](#side-database)
+
+## Implementations
+
+- JavaScript
+  - Data Formats - [unixfs](https://github.com/ipfs/js-ipfs-unixfs)
+  - Importer - [unixfs-importer](https://github.com/ipfs/js-ipfs-unixfs-importer)
+  - Exporter - [unixfs-exporter](https://github.com/ipfs/js-ipfs-unixfs-exporter)
+- Go
+  - [`ipfs/go-ipfs/unixfs`](https://github.com/ipfs/go-ipfs/tree/b3faaad1310bcc32dc3dd24e1919e9edf51edba8/unixfs)
+  - Protocol Buffer Definitions - [`ipfs/go-ipfs/unixfs/pb`](https://github.com/ipfs/go-ipfs/blob/b3faaad1310bcc32dc3dd24e1919e9edf51edba8/unixfs/pb/unixfs.proto)
+
+## Data Format
+
+The UnixfsV1 data format is represented by this protobuf:
+
+```protobuf
+message Data {
+	enum DataType {
+		Raw = 0;
```

Comment:
I belive `Raw` (`Data.DataType`) is unused (I didn't even noticed it).

I guess it's used by IPLD reiffiers to indicate `0x55`.

Reply:
Investigating it seems like this is an implementation detail of kubo that leaked out (for some reason kubo internally differenciate files and raw-files, this is a distinction futures implementations should avoid as this create more edge cases to deal with (I would rather say that `File` is an `interface` or `trait` that is both implemeted by `Raw` blocks and `Dag-PB` ones with `PBNode.Data.Type == File`)).

```console
$ rgrep -I "\bRaw\b" | grep "\.go:" | grep unixfs
vendor/github.com/ipfs/go-unixfsnode/data/datatypes.go:	Data_Raw:       "Raw",
vendor/github.com/ipfs/go-unixfsnode/data/datatypes.go:	"Raw":       Data_Raw,
vendor/github.com/ipfs/go-unixfs/unixfs.go:// Raw nodes are (also) processed because they are used as leaf
vendor/github.com/ipfs/go-unixfs/importer/trickle/trickledag.go:			codec = cid.Raw
vendor/github.com/ipfs/go-unixfs/importer/trickle/trickledag.go:		if codec == cid.Raw && expect.Version == 0 {
vendor/github.com/ipfs/go-unixfs/importer/helpers/dagbuilder.go:// the UnixFS layer node (either `File` or `Raw`).
vendor/github.com/ipfs/go-unixfs/pb/unixfs.pb.go:	0: "Raw",
vendor/github.com/ipfs/go-unixfs/pb/unixfs.pb.go:	"Raw":       0,
core/coreapi/unixfs.go:	case cid.Raw:
```

I'll add a comment about this field being reserved and remove it.

Reply:
TRaw is actually used in some places, and existed before IPLD (and therefore the 0x55 codec). A small amount of poking around shows that the use of TFile for leaf nodes was a bug that was kept around, but that TRaw was also in use and so also couldn't be dropped.

https://github.com/ipfs/kubo/pull/5120

TRaw doesn't seem any worse than TFile for leaf nodes, unless you're trying to recreate old imports (generally a bad idea where you'll run into rough edges) or absolutely need CIDv0 links everywhere (including inside the DAG) using the `raw` (0x55) codec is optimal.